pkg/ddl/job_table.go

```diff
@@ -526,10 +523,15 @@ func (s *jobScheduler) delivery2Worker(wk *worker, pool *workerPool, job *model.
 	jobID, involvedSchemaInfos := job.ID, job.GetInvolvingSchemaInfo()
 	s.runningJobs.addRunning(jobID, involvedSchemaInfos)
 	metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Inc()
-	s.wg.RunWithLog(func() {
+	s.wg.Run(func() {
 		defer func() {
+			r := recover()
+			if r != nil {
+				logutil.DDLLogger().Error("panic in delivery2Worker", zap.Any("recover", r), zap.Stack("stack"))
+			}
 			failpoint.InjectCall("afterDelivery2Worker", job)
-			s.runningJobs.removeRunning(jobID, involvedSchemaInfos)
+			moveRunningJobsToPending := r != nil || (job != nil && job.IsPaused())
```

Comment:
how about when `delivery2Worker` exits, if the job is not in final states, we move it to pending.

Reply:
https://github.com/pingcap/tidb/pull/54419/commits/7820b1f9906a583c40a8f2300158cc687e021d52

Reply:
if not panic/owner change, pause is the only state if we exit here actually, see `job.InFinalState()` branch

Reply:
just in case we will have more states in future (for example, somehow we revert to the old way to handle job, to let this function only forward one schema state). I think checking "not finished" is more correct.

Reply:
lgtm

> more correct

you mean 'more robust'?

Reply:
yes, more robust.